spring-cloud-loadbalancer/src/test/java/org/springframework/cloud/loadbalancer/core/WeightedServiceInstanceListSupplierTest.java

```diff
@@ -0,0 +1,236 @@
+		Map<String, Integer> counter = new HashMap<>();
+		for (ServiceInstance serviceInstance : serviceInstances) {
+			counter.compute(serviceInstance.getInstanceId(), (k, v) -> {
+				if (v == null) {
+					return 1;
+				}
+				return v + 1;
+			});
+		}
+		assertThat(counter).containsEntry("test-1", DEFAULT_WEIGHT);
+		assertThat(counter).containsEntry("test-2", DEFAULT_WEIGHT);
+		assertThat(counter).containsEntry("test-3", 3);
+	}
+
+	@Test
+	void shouldUseDefaultWeightWhenWeightIsNotNumber() {
+		var one = serviceInstance("test-1", buildWeightMetadata("Foo"));
+		var two = serviceInstance("test-2", buildWeightMetadata("Bar"));
+		var three = serviceInstance("test-3", buildWeightMetadata("Baz"));
+
+		when(delegate.get()).thenReturn(Flux.just(Arrays.asList(one, two, three)));
+		var supplier = new WeightedServiceInstanceListSupplier(delegate);
+
+		List<ServiceInstance> serviceInstances = Objects.requireNonNull(supplier.get().blockFirst());
+		Map<String, Integer> counter = new HashMap<>();
+		for (ServiceInstance serviceInstance : serviceInstances) {
+			counter.compute(serviceInstance.getInstanceId(), (k, v) -> {
+				if (v == null) {
+					return 1;
+				}
+				return v + 1;
+			});
+		}
+		assertThat(counter).containsEntry("test-1", DEFAULT_WEIGHT);
+		assertThat(counter).containsEntry("test-2", DEFAULT_WEIGHT);
+		assertThat(counter).containsEntry("test-3", DEFAULT_WEIGHT);
+	}
+
+	@Test
+	void shouldUseDefaultWeightWhenWeightedFunctionReturnZero() {
+		var one = serviceInstance("test-1", Collections.emptyMap());
+		var two = serviceInstance("test-2", Collections.emptyMap());
+		var three = serviceInstance("test-3", Collections.emptyMap());
+
+		when(delegate.get()).thenReturn(Flux.just(Arrays.asList(one, two, three)));
+		var supplier = new WeightedServiceInstanceListSupplier(delegate, instance -> 0);
+
+		List<ServiceInstance> serviceInstances = Objects.requireNonNull(supplier.get().blockFirst());
+		Map<String, Integer> counter = new HashMap<>();
+		for (ServiceInstance serviceInstance : serviceInstances) {
+			counter.compute(serviceInstance.getInstanceId(), (k, v) -> {
+				if (v == null) {
+					return 1;
+				}
+				return v + 1;
+			});
+		}
+		assertThat(counter).containsEntry("test-1", DEFAULT_WEIGHT);
+		assertThat(counter).containsEntry("test-2", DEFAULT_WEIGHT);
+		assertThat(counter).containsEntry("test-3", DEFAULT_WEIGHT);
+	}
+
+	@Test
+	void shouldUseDefaultWeightWhenWeightedFunctionReturnNegative() {
+		var one = serviceInstance("test-1", Collections.emptyMap());
+		var two = serviceInstance("test-2", Collections.emptyMap());
+		var three = serviceInstance("test-3", Collections.emptyMap());
+
+		when(delegate.get()).thenReturn(Flux.just(Arrays.asList(one, two, three)));
+		var supplier = new WeightedServiceInstanceListSupplier(delegate, instance -> -1);
+
+		List<ServiceInstance> serviceInstances = Objects.requireNonNull(supplier.get().blockFirst());
+		Map<String, Integer> counter = new HashMap<>();
```

Comment:
This counter repeats a lot. Maybe you could extract it to a method and reuse in various tests?